Fix assertion failure in gemm template lowering

dmpots opened the pull request:
Summary:
This commit fixes a crash in the gemm template lowering caused by hitting an [assert](https://github.com/pytorch/pytorch/blob/fd515e4f59bfa0ac9faa5185b7a02f3222c4cd08/torch/_inductor/codegen/common.py#L1181) that a buffer was previously removed.

The assert triggers because in the first gemm lowering we use a local accumulation buffer, which causes the original buffer name to be added to the `removed_buffers` set. Then in the next gemm lowering we use the global buffer for accumulation, but that buffer name is already in the `removed_buffers` set.

The fix is to add a unique suffix to the buffer name to avoid triggering the assert from different gemm lowerings.


Differential Revision: D68814625




cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @amjames @desertfire @chauhang @aakhundov

dmpots commented:
@pytorchbot label "topic: not user facing"

hl475 commented:
@leslie-fang-intel and @frost-intel - could you please help take a look? we found another issue when trying autuotune from Meta internal models, and David comes up with this fix

leslie-fang-intel commented on test/inductor/test_cpu_select_algorithm.py:
```diff
@@ -2299,6 +2299,53 @@ def forward(self, x, w):
+            def forward(self, x):
+                y = self.linear(x)
+                view = torch.ops.aten.view.default(y, [-1, in_features1])
+                return self.linear1(view)
```
> The assert triggers because in the first gemm lowering we use a local accumulation buffer, which causes the original buffer name to be added to the removed_buffers set. Then in the next gemm lowering we use the global buffer for accumulation, but that buffer name is already in the removed_buffers set

Thanks for the PR. Feels like in this testcase the first and second gemm's output should be corresponding to 2 different output buffers and shouldn't hit the case descripted in the summary. Could you kindly elaborate more about the how the issue happens with this test case?

dmpots replied:
This test case was reduced from a larger model. I may be using some of the wrong terminology in the description so let me try to explain in more detail.

The two operators here are intended to induce different values for `use_local_acc`

https://github.com/pytorch/pytorch/blob/23fffb54d591c7b5ca6d19728d628dbe1e79d91c/torch/_inductor/codegen/cpp_gemm_template.py#L1177

In the lowering of the first operator (linear) `use_local_acc == True`. This causes the following check to be true

https://github.com/pytorch/pytorch/blob/157d81c201715f84ead21d0ee420669ab7f58c04/torch/_inductor/codegen/cpp_template_kernel.py#L371

since we have 
```
orig_src.get_name() == "buf_out_GemmOut"
src.get_name() == "local_acc_buf"
```

Which causes a call to `add_local_buffer`. In there the "buf_out_GemmOut" is added to the graph's `removed_buffers` set.
                
https://github.com/pytorch/pytorch/blob/157d81c201715f84ead21d0ee420669ab7f58c04/torch/_inductor/codegen/cpp_utils.py#L431

In the lowering of the second operator (linear1) `use_local_acc == False`, which causes the same check to be false

https://github.com/pytorch/pytorch/blob/157d81c201715f84ead21d0ee420669ab7f58c04/torch/_inductor/codegen/cpp_template_kernel.py#L371

since we have 
```
orig_src.get_name() == "buf_out_GemmOut"
src.get_name() == "buf_out_GemmOut"
```
This means the buffer is not added as a local buffer. So when it goes to generate a call to the microkernel here

https://github.com/pytorch/pytorch/blob/7f65a208848205b38445423b7e2e93a2b4994e5e/torch/_inductor/codegen/cpp_micro_gemm.py#L145

we fall into the `else` branch to lookup the value from `self.args.args` instead of `self.local_buffers`

https://github.com/pytorch/pytorch/blob/7f65a208848205b38445423b7e2e93a2b4994e5e/torch/_inductor/codegen/cpp_template_kernel.py#L148

and that is where we hit the assert that the buffer name (`buf_out_GemmOut`) is already in the `removed_buffers` set

https://github.com/pytorch/pytorch/blob/fd515e4f59bfa0ac9faa5185b7a02f3222c4cd08/torch/_inductor/codegen/common.py#L1181

Hopefully, that makes more sense now. Curious to hear if you have a better suggestion to fix the issue.

frost-intel replied:
I wasn't able to reproduce the error with your test as written? But if I changed `out_features1 = 32`, then I saw the error.

dmpots replied:
Oh, interesting. It does seem to repro on my end with `out_features1 = 24`. I think the place that size matters is the check here when setting `use_local_acc`

https://github.com/pytorch/pytorch/blob/23fffb54d591c7b5ca6d19728d628dbe1e79d91c/torch/_inductor/codegen/cpp_gemm_template.py#L1181

Does the blocking size depend on the machine executing? In my case the `padded_n` and `n` values are both 24 when I run in the debugger.

frost-intel replied:
Yes, it's dependent on the vectorization architecture, specified here: https://github.com/pytorch/pytorch/blob/fd0cd6a08f706b7bb1dedb296217b6441e4fb9ff/torch/_inductor/codegen/cpp_micro_gemm.py#L283-L287
vs 
https://github.com/pytorch/pytorch/blob/fd0cd6a08f706b7bb1dedb296217b6441e4fb9ff/torch/_inductor/codegen/cpp_micro_gemm.py#L308-L312
Both options are submitted for consideration, but the blocking will likely be more efficient if AVX512 is available.

Solution should be to parameterize and run for both 24 and 32.

leslie-fang-intel replied:
> But if I changed out_features1 = 32, then I saw the error.

Change it to this size, I still can't reproduce this failure. Maybe we can fix the number of cores and fix the `config.cpp.gemm_cache_blocking` to reproduce this failure steadily on different system.

dmpots replied:
I pushed a change to try the different blocking sizes `(8,16,24,32,48,)` I see in the `generate_gemm_config` calls.

I wasn't sure how to use `config.cpp.gemm_cache_blocking` here since it looks like the padded_n value is computed based on `thread_blocking` instead of `cache_blocking`.

leslie-fang-intel replied:
`config.cpp.gemm_thread_factors` can be used to decide the `thread_blocking`. This test case https://github.com/pytorch/pytorch/blob/fa0592b568ce35c391dfede8bede9e1ed67f33f5/test/inductor/test_cpu_select_algorithm.py#L1691 might be a example for how to use it.

dmpots commented on torch/_inductor/codegen/cpp_gemm_template.py:
```diff
@@ -1253,7 +1254,10 @@ def copy_inner(index):
+            gemm_output_name = (
+                f"{template_buffer.get_name()}_GemmOut" + f"_{name_suffix}"
+            )
             gemm_output_buffer = ir.Buffer(
```
That looks like a better way to me. Updated the PR with this suggestion. Thanks!

dmpots committed:
Fix assertion failure in gemm template lowering (#146353)

Summary:

This commit fixes a crash in the gemm template lowering caused by hitting an [assert](https://github.com/pytorch/pytorch/blob/fd515e4f59bfa0ac9faa5185b7a02f3222c4cd08/torch/_inductor/codegen/common.py#L1181) that a buffer was previously removed.

The assert triggers because in the first gemm lowering we use a local accumulation buffer, which causes the original buffer name to be added to the `removed_buffers` set. Then in the next gemm lowering we use the global buffer for accumulation, but that buffer name is already in the `removed_buffers` set.

The fix is to add a unique suffix to the buffer name to avoid triggering the assert from different gemm lowerings.

Test Plan:
# Reduced test case
```
TRITON_LOCAL_BUILD=1 buck2 run 'fbcode//mode/opt' fbcode//caffe2/test/inductor:cpu_select_algorithm_cpu  -- caffe2.test.inductor.test_cpu_select_algorithm.TestSelectAlgorithmCPU.test_local_and_global_accumulator_cpu_float32
```

# Original failure
```
$ manifold get aitemplate/tree/aotinductor_cpu/915857944_1.input.predictor.disagg.remote_other /tmp/915857944_1.input.predictor.disagg.remote_other

$ buck2 run @//mode/opt //deeplearning/aot_inductor/cpu:cli -- --local-model-path /tmp/915857944_1.input.predictor.disagg.remote_other --submodule remote_other --preset ads_second_stage_ranking_type_1
```

Differential Revision: D68814625

hl475 commented:
@pytorchbot merge

hl475 commented:
@pytorchbot merge -f "the failed test is irrelevant with this PR"
